functions/src/stories.ts

```diff
@@ -11,83 +11,161 @@ const index = algoliasearch(
+          .get()
+          .then((doc) => doc.data()!);
+
+      const entry: StoryRecord = {
+        storyID,
+        storyAuthorID: story.metaData.authorId,
+        translationID,
+        translationAuthorID: translation.metaData.authorId,
+        language: translation.language,
+        title,
+        description,
+        tags,
+      };
       if (change.before.exists) {
-        await index.deleteBy({
-          filters: "translationID:" + translationID,
-        });
-      }
-      if (change.after.exists) {
-        const {title, description, tags} = change.after.data()!;
-        const story = await db
-            .doc(`stories/${storyID}`)
-            .get()
-            .then((doc) => doc.data()!);
-        const translation = await db
-            .doc(`stories/${storyID}/translations/${translationID}`)
-            .get()
-            .then((doc) => doc.data()!);
+        await index.partialUpdateObject(
+            {objectID: translationID, ...entry},
+            {createIfNotExists: true}
+        );
+      } else {
         await index.saveObject(
-        {
-          storyID,
-          storyAuthorID: story.metaData.authorId,
-          translationID,
-          translationAuthorID: translation.metaData.authorId,
-          language: translation.language,
-          title,
-          description,
-          tags,
-          likes: translation.metaData.likes,
-        } as storyRecord,
-        {autoGenerateObjectIDIfNotExist: true}
+            entry,
+            {autoGenerateObjectIDIfNotExist: true}
         );
       }
     });
 
-// TODO
+export const trackStoryUpdateTime = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}/assets/{assetID}"
+    )
+    .onWrite(async (change, context) => {
+      if (!change.after.exists) return;
+      await storyDoc(context.params.storyID, context.params.translationID)
+          .update({
+            "metaData.lastUpdateAt": firestore.FieldValue.serverTimestamp(),
+          });
+    });
+
+export const updateStoryMeta = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}"
+    )
+    .onWrite(async (change, context) => {
+      if (change.after.exists) {
+        const meta = change.after.data()?.metaData;
+        await index.partialUpdateObject({
+          objectID: context.params.translationID,
+          likes: meta.likes ?? 0,
+          views: meta.views ?? 0,
+          lastUpdateAt: meta.lastUpdateAt,
+          imageUrl: meta.imageUrl,
+        });
+      }
+    });
+
+export const indexTrendingStories = functions
+    .region("europe-central2")
+    .pubsub.schedule("every 72 hours")
+    .onRun(async () => {
+      await index.browseObjects<StoryRecord>({
+        query: "",
+        attributesToRetrieve: ["storyID"],
+        batch: async (batch) => {
+          for (const hit of batch) {
```

Comment:
I don't quite understand what's going on. Is it fetching all the stories from Algolia, getting just their IDs, then using the ID to get the document itself from Firestore, and then checking the metadata that was added using Algolia previously? Why not query the Firestore collection directly?

Reply:
I was thinking to perform this only on Algolia entries to omit private contents. But then again we can do all of that using multi-collection queries. Yes.